memtagabielf64/memtagabielf64.rst

```diff
@@ -0,0 +1,723 @@
+      last_symbol_end = symbol.address + symbol.size
+
+      if size < (1 << distance_reserved_bits):
+         write_to_section(uleb128(distance | size))
+      else:
+         write_to_section(uleb128(distance))
+         write_to_section(uleb128(size - 1))
+
+.. raw:: pdf
+
+   PageBreak
+
+Relocation Operations
+---------------------
+
+In order to support materialization of the address tag into the relevant GOT
+entry for a tagged global, some AArch64 relocations have semantics that extend
+the AAELF64_ definition. Because global variable support in the MemtagABI is
+designed to be backwards-compatible, the result of the relocation on a
+non-MemtagABI binary or non-MTE device is the same as the non-extended variant.
+
+The relocations reference the following mnemonics:
+
+- ``LDG(value)`` is an instruction for the run-time environment to materialize a
+  tagged pointer from the provided address, using the ``ldg`` instruction. This
+  operation should also align the pointer down to the closest tag granule before
+  executing the ``ldg`` instruction.
+
+- ``TAG(pointer, offset)`` is an instruction for the run-time environment to
+  fetch the memory tag at ``(pointer + offset) & ~0xf``, which can be done using
+  the ``ldg`` instruction. Then, this tag should be materialized to ``pointer``,
+  overwriting the existing address tag in ``pointer`` if it exists.
+
+- For all the relocation types listed below, the loader should use the ``ldg``
+  instruction on the target address before writing into the target field, as the
+  target field may be inside of a tagged region.
+
+.. table:: Relocations with extended semantics
+
+  +------------+--------------------+----------------+-------------------------------------+
+  | ELF64 Code | Name               | Base Operation | MemtagABI Extended Operation        |
+  +============+====================+================+=====================================+
+  | 257	      | R_AARCH64_ABS64    | S + A          | LDG(S) + A                          |
+  +------------+--------------------+----------------+-------------------------------------+
+  | 1025	      | R_AARCH64_GLOB_DAT | S + A          | LDG(S) + A                          |
+  +------------+--------------------+----------------+-------------------------------------+
+  | 1027	      | R_AARCH64_RELATIVE | Delta(S) + A   | TAG(Delta(S) + A, \*(Delta(S) + A)) |
+  +------------+--------------------+----------------+-------------------------------------+
+
+``R_AARCH64_ABS64`` and ``R_AARCH64_GLOB_DAT`` are thus extended to materialize
+the memory tag of the symbol ``S`` for relocation purposes.
+
+Extended semantics of ``R_AARCH64_RELATIVE``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+``R_AARCH64_RELATIVE`` is extended with special care to preserve backwards
+compatibility. This relocation extension allows for the out-of-bounds derived
+symbols to be resolved with the correct memory tags, as shown in the example
+below:
```

Comment:
Might be good to include the specific term "tag derivation" in this first paragraph because later on you use it and it's not very clear that these examples are examples of "tag derivation". You use almost the same words in this intro but not quite.

Reply:
thanks. hopefully i've clearly defined it below (in the comment about `GOT_ENTRY_FOR_foo_end`).

Reply:
For whatever reason, I can't resolve my own comment, but this looks good to me.